### ISSUE TITLE
Time series forecasting / prediction

### ISSUE THREAD

**Issue by amotl** [daq-tools/kotori#168]
## About
At [^1][^2], we shared a few notes about time series anomaly detection, and forecasting/prediction. Other than using traditional time series forecasting libraries like [Prophet](https://github.com/facebook/prophet) and friends, there are now also other kinds of prediction methods, based on machine learning models, like [TimeGPT](https://towardsdatascience.com/timegpt-the-first-foundation-model-for-time-series-forecasting-bf0a75e63b3a), or [Chronos](https://github.com/amazon-science/chronos-forecasting), that allow for zero-shot inference.

[^1]: https://kotori.readthedocs.io/en/latest/development/research/timeseries-analysis.html
[^2]: https://kotori.readthedocs.io/en/latest/development/backlog.html

**Comment by amotl** [comment]
## ATFNet

> Adaptive Time-Frequency Ensembled Network for Long-term Time Series Forecasting.

https://github.com/YHYHYHYHYHY/ATFNet

**Comment by amotl** [comment]
## EarthPT

> A simple repository for training time series large observation models. This repository began its life as Andrej Karpathy's [nanoGPT](https://github.com/karpathy/nanoGPT), and has been altered so that it is usable for time series data.

https://github.com/aspiaspace/earthPT

**Comment by amotl** [comment]
## TinyTimeMixer (TTM)

### About
> TinyTimeMixers (TTMs) are compact pre-trained models for Multivariate Time-Series Forecasting, open-sourced by IBM Research. With less than 1 Million parameters, TTM introduces the notion of the first-ever “tiny” pre-trained models for Time-Series Forecasting.
>
> The current open-source version supports point forecasting use-cases ranging from minutely to hourly resolutions (Ex. 10 min, 15 min, 1 hour, etc.). Note that zeroshot, fine-tuning and inference tasks using TTM can easily be executed in 1 GPU machine or in laptops too.

### Details
TTM-1 currently supports 2 modes:
- **Zeroshot forecasting:** Directly apply the pre-trained model on your target data to get an initial forecast (with no training).
- **Finetuned forecasting:** Finetune the pre-trained model with a subset of your target data to further improve the forecast.

> Since, TTM models are extremely small and fast, it is practically very easy to finetune the model with your available target data in few minutes to get more accurate forecasts. For more details on TTM architecture and benchmarks, refer to our [paper](https://arxiv.org/pdf/2401.03955.pdf).

https://huggingface.co/ibm/TTM

**Comment by amotl** [comment]
## Neural 🧠 Forecast

> **NeuralForecast** offers a large collection of neural forecasting models focused on their usability, and robustness. The models range from classic networks like `MLP`, `RNN`s to novel proven contributions like `NBEATS`, `NHITS`, `TFT` and other architectures.

#### 🎊 Features

- Exogenous Variables: Static, historic and future exogenous support.
- Forecast Interpretability: Plot trend, seasonality and exogenous [NBEATS](https://nixtla.github.io/neuralforecast/models.nbeats.html#nbeats), [NHITS](https://nixtla.github.io/neuralforecast/models.nhits.html#nhits), [TFT](https://nixtla.github.io/neuralforecast/models.tft.html#tft), ESRNN prediction components.
- Probabilistic Forecasting: Simple model adapters for quantile losses and parametric distributions.
- Train and Evaluation Losses Scale-dependent, percentage and scale independent errors, and parametric likelihoods.
- Automatic Model Selection Parallelized automatic hyperparameter tuning, that efficiently searches best validation configuration.
- Simple Interface Unified SKLearn Interface for StatsForecast and MLForecast compatibility.
- Model Collection: Out of the box implementation of [MLP](https://nixtla.github.io/neuralforecast/models.mlp.html#mlp), [LSTM](https://nixtla.github.io/neuralforecast/models.lstm.html#lstm), [RNN](https://nixtla.github.io/neuralforecast/models.rnn.html#rnn), [TCN](https://nixtla.github.io/neuralforecast/models.tcn.html#tcn), [DilatedRNN](https://nixtla.github.io/neuralforecast/models.dilated_rnn.html#dilatedrnn), [NBEATS](https://nixtla.github.io/neuralforecast/models.nbeats.html#nbeats), [NHITS](https://nixtla.github.io/neuralforecast/models.nhits.html#nhits), ESRNN, [Informer](https://nixtla.github.io/neuralforecast/models.informer.html#informer), [TFT](https://nixtla.github.io/neuralforecast/models.tft.html#tft), [PatchTST](https://nixtla.github.io/neuralforecast/models.patchtst.html#patchtst), [VanillaTransformer](https://nixtla.github.io/neuralforecast/models.vanillatransformer.html#vanillatransformer), [StemGNN](https://nixtla.github.io/neuralforecast/models.stemgnn.html#stemgnn) and [HINT](https://nixtla.github.io/neuralforecast/models.hint.html#hint). See the entire [collection here](https://nixtla.github.io/neuralforecast/models.html).

Web: https://nixtlaverse.nixtla.io/neuralforecast/
Repository: https://github.com/Nixtla/neuralforecast

**Comment by amotl** [comment]
## Awesome AI for Time Series (AI4TS) Papers, Tutorials, and Surveys

> A professionally curated list of papers (with available code), tutorials, and surveys on recent AI for Time Series Analysis (AI4TS), including Time Series, Spatio-Temporal Data, Event Data, Sequence Data, Temporal Point Processes, etc., at the Top AI Conferences and Journals, which is updated ASAP (the earliest time) once the accepted papers are announced in the corresponding top AI conferences/journals. Hope this list would be helpful for researchers and engineers who are interested in AI for Time Series Analysis.

Repository: https://github.com/qingsongedu/awesome-AI-for-time-series-papers

**Comment by amotl** [comment]
## PatchTST (ICLR 2023)

> We propose an efficient design of Transformer-based models for multivariate time series forecasting and self-supervised representation learning. It is based on two key components: (i) segmentation of time series into subseries-level patches which are served as input tokens to Transformer; (ii) channel-independence where each channel contains a single univariate time series that shares the same embedding and Transformer weights across all the series. Patching design naturally has three-fold benefit: local semantic information is retained in the embedding; computation and memory usage of the attention maps are quadratically reduced given the same look-back window; and the model can attend longer history.
>
> Our channel-independent patch time series Transformer (PatchTST) can improve the long-term forecasting accuracy significantly when compared with that of SOTA Transformer-based models. We also apply our model to self-supervised pre-training tasks and attain excellent fine-tuning performance, which outperforms supervised training on large datasets. Transferring of masked pre-trained representation on one dataset to others also produces SOTA forecasting accuracy.

Paper: [A Time Series is Worth 64 Words: Long-term Forecasting with Transformers](https://arxiv.org/abs/2211.14730)
Repository: https://github.com/yuqinie98/PatchTST

**Comment by amotl** [comment]
## tsai

_State-of-the-art Deep Learning library for Time Series and Sequences._

> `tsai` is an open-source deep learning package built on top of Pytorch & fastai focused on state-of-the-art techniques for time series tasks like classification, regression, forecasting, imputation…
>
> `tsai` is currently under active development by timeseriesAI.

Repository: https://github.com/timeseriesAI/tsai
Documentation: https://timeseriesai.github.io/tsai/

**Comment by amotl** [comment]
## skforecast

> Time series forecasting with scikit-learn models.

> Skforecast is a Python library that eases using scikit-learn regressors as single and multi-step forecasters. It also works with any regressor compatible with the scikit-learn API (LightGBM, XGBoost, CatBoost, ...).

Homepage: https://skforecast.org/
Repository: https://github.com/JoaquinAmatRodrigo/skforecast

**Comment by amotl** [comment]
## TabPFN

_Accurate predictions on small data with a tabular foundation model._
- https://news.ycombinator.com/item?id=42647343
- https://www.nature.com/articles/s41586-024-08328-6
- https://github.com/PriorLabs/tabpfn

_Implementation of TabPFN-TS (work accepted at NeurIPS 2024 TRL and TSALM workshops)._
The Tabular Foundation Model TabPFN Outperforms Specialized Time Series Forecasting Models Based on Simple Features.
- https://github.com/liam-sbhoo/tabpfn-time-series
- https://arxiv.org/abs/2501.02945

**Comment by amotl** [comment]
## Grafana Augurs

> This repository contains `augurs`, a time series toolkit built in Rust. It aims to provide some useful primitives for working with time series, as well as the main functionality: heavily optimized models for forecasting, outlier detection, clustering, seasonality detection, changepoint detection and more. Most algorithms are based on existing R or Python implementations.

https://github.com/grafana/augurs

**Comment by amotl** [comment]
## PyPOTS

A Python toolbox for machine learning on Partially-Observed Time Series.

### About

A Python toolkit/library for reality-centric machine/deep learning and data mining on partially-observed time series, including SOTA neural network models for scientific analysis tasks of imputation/classification/clustering/forecasting/anomaly detection/cleaning on incomplete industrial (irregularly-sampled) multivariate TS with NaN missing values.

### Resources
- https://pypots.com/
- https://github.com/WenjieDu/PyPOTS

**Comment by amotl** [comment]
## Darts

Darts is a Python library for user-friendly forecasting and anomaly detection on time series. It contains a variety of models, from classics such as ARIMA to deep neural networks.

- https://github.com/unit8co/darts
- https://unit8co.github.io/darts/

**Comment by amotl** [comment]
## Python Outlier Detection (PyOD) V2

> A Python Library for Outlier and Anomaly Detection, Integrating Classical and Deep Learning Techniques 

-- https://github.com/yzhao062/pyod
-- https://pyod.readthedocs.io/

**Comment by amotl** [comment]
## IBM Granite

IBM Granite provides a few models, one is also about time series.

- https://www.ibm.com/granite
- https://huggingface.co/collections/ibm-granite/granite-time-series

**Comment by amotl** [comment]
## STUMPY

STUMPY is a powerful and scalable Python library for modern time series analysis.

- https://github.com/stumpy-dev/stumpy
- https://stumpy.readthedocs.io/

**Comment by amotl** [comment]
## ADS

The Oracle Accelerated Data Science (ADS) package seems to bundle a few gems.

https://accelerated-data-science.readthedocs.io/

**Comment by amotl** [comment]
## tsfm-onnx

[tsfm-onnx](https://github.com/sktime/tsfm-onnx) wraps [Chronos-2](https://huggingface.co/amazon/chronos-2), [t0-alpha](https://huggingface.co/theforecastingcompany/t0-alpha), [Toto-2.0-22m](https://huggingface.co/Datadog/Toto-2.0-22m), [TinyCast](https://huggingface.co/raws-labs/tinycast), and [TinyTimeMixer r2](https://huggingface.co/ibm-granite/granite-timeseries-ttm-r2) models by pushing the boundaries to even make them available in your browser, see @fkiraly's article [The Temporer Has No Clothes](https://www.sktime.net/blog/forecasting-in-the-browser/). Because it is possible to also [use the models from Python](https://github.com/sktime/tsfm-onnx#use-the-models-from-python), the package might also be a good candidate to pull into Kotori DAQ?

Cheers, @siddharth7113 and @closechoice.
